### ISSUE TITLE
`month`, `yr` and `dec` durations are broken

### ISSUE THREAD

**Issue by amtoine** [nushell/nushell#8036]
### Describe the bug

given what `into duration` does, i think `yr`, `month` and `dec` should be part of the "duration set", but it looks like `nushell` does not recognize them :thinking: 

## :heavy_check_mark: the following work fine
```bash
> [1ns 1us 1ms 1sec 1min 1hr 1day 1wk]
╭───┬──────╮
│ 0 │  1ns │
│ 1 │  1µs │
│ 2 │  1ms │
│ 3 │ 1sec │
│ 4 │ 1min │
│ 5 │  1hr │
│ 6 │ 1day │
│ 7 │  1wk │
╰───┴──────╯
```

## :x: but these do not work...
```bash
> "1month" | into duration
Error: nu::shell::cant_convert_with_value (link)

  × Can't convert string `1month` to duration.
   ╭─[entry #53:1:1]
 1 │ "1month" | into duration
   · ────┬───   ──────┬──────
   ·     │            ╰── can't be converted to duration
   ·     ╰── this string value...
   ╰────
  help: supported units are ns, us, ms, sec, min, hr, day, wk, month, yr and dec
```
```bash
> "1yr" | into duration
Error: nu::shell::cant_convert_with_value (link)

  × Can't convert string `1yr` to duration.
   ╭─[entry #54:1:1]
 1 │ "1yr" | into duration
   · ──┬──   ──────┬──────
   ·   │           ╰── can't be converted to duration
   ·   ╰── this string value...
   ╰────
  help: supported units are ns, us, ms, sec, min, hr, day, wk, month, yr and dec
```
```bash
> "1dec" | into duration
Error: nu::shell::cant_convert_with_value (link)

  × Can't convert string `1dec` to duration.
   ╭─[entry #55:1:1]
 1 │ "1dec" | into duration
   · ───┬──   ──────┬──────
   ·    │           ╰── can't be converted to duration
   ·    ╰── this string value...
   ╰────
  help: supported units are ns, us, ms, sec, min, hr, day, wk, month, yr and dec
```

> **Note**
> what's really strange and looks like a bug, either in the doc or in the code, is that `nushell` says
> > help: supported units are ns, us, ms, sec, min, hr, day, wk, month, yr and dec

### How to reproduce

1. `[1ns 1us 1ms 1sec 1min 1hr 1day 1wk]` works
2. `[1ns 1us 1ms 1sec 1min 1hr 1day 1wk] | into duration` works
3. `"1month" | into duration` gives a `nu::shell::cant_convert_with_value` error
4. `"1yr" | into duration` gives a `nu::shell::cant_convert_with_value` error
5. `"1dec" | into duration` gives a `nu::shell::cant_convert_with_value` error

### Expected behavior

depending on where the bug really comes from
- remove "month", "yr" and "dec" from the help message in the `nu::shell::cant_convert_with_value` error
- add support for `month`, `yr` and `dec` back into `into duration`

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.75.1                                   |
| branch             | main                                     |
| commit_hash        | f4bd78b86d441cd53fb6ff30d3a6423735a423cc |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.66.1 (90743e729 2023-01-10)      |
| rust_channel       | 1.66.1-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.66.1 (ad779e08b 2023-01-10)      |
| pkg_version        | 0.75.1                                   |
| build_time         | 2023-02-11 11:32:11 +01:00               |
| build_rust_channel | release                                  |
| features           | database, default, trash, which, zip     |
| installed_plugins  |                                          |

**Comment by fdncred** [comment]
They have been added, removed, added, removed. I think they should stay but others think that since all months don't have the same number of days and that there are leap years, that they shouldn't be included.

The reason they're partially there now is because we liked the way the banner enumerates the time since nushell's first commit.

**Comment by bobhy** [comment]
Agreed, a duration result is not currently always round-trippable, and that, for me is good justification to add the longer durations back.
```
〉("28-feb-2022" | into datetime) - ("28-feb-2000" | into datetime)
22yr 6day
------------------------------------------------------------
〉("28-feb-2000" | into datetime) + 22yr + 6day
Error: nu::shell::type_mismatch (link)

  × Type mismatch during operation.
   ╭─[entry #101:1:1]
 1 │ ("28-feb-2000" | into datetime) + 22yr + 6day
   ·  ──────┬──────                  ┬ ──┬─
   ·        │                        │   ╰── string
   ·        │                        ╰── type mismatch for operator
   ·        ╰── date
   ╰────
```

**Comment by amtoine** [comment]
> They have been added, removed, added, removed. I think they should stay but others think that since all months don't have the same number of days and that there are leap years, that they shouldn't be included.
> 
> The reason they're partially there now is because we liked the way the banner enumerates the time since nushell's first commit.
>
oh i see :+1: 

but from a user perspective, without knowing this, it's quite confusing :wink: 

then maybe this could be discussed here? :relieved:

**Comment by fdncred** [comment]
There are a couple points to consider.
1. durations are stored as nanoseconds
2. date math is a different thing

Here's really what those mean. Any nanosecond can be broken down into a nushell duration that pretty printed. Currently date math will return a duration, which converted into a pretty print can be wrong. 

Durations no nothing about leap years and that's the problem. In order to do date math, you really need to know the start date and the end date and not only a bunch of nanoseconds.

In bob's example, rewritten more simply, you can do this in nushell
```
2022-02-28 - 2000-02-28
```
and it gives the same answer `22yr 6day` but that answer is actually wrong. We know that answer to be `22yr` or 8036 days. You can see that answer here too https://www.timeanddate.com/date/durationresult.html?m1=2&d1=28&y1=2000&m2=2&d2=28&y2=2022

Interestingly enough, our chrono crate understand all the leap year stuff. You can do this and get the right answer in days.
```
> 2022-02-28 - 2000-02-28 | into duration --convert day
8036
```

You can also see here, that the math, to the nanosecond, is correct here.
```
> 2022-02-28 - 2000-02-28 | debug -r
Duration {
    val: 694310400000000000,
    span: Span {
        start: 110906,
        end: 110929,
    },
}
```
because that website above says the number of seconds is `694,310,400 seconds` which is correct because we can do
```
> 694_310_400 * 1_000_000_000
694310400000000000
```
So, what went wrong? The problem is what I said earlier, with date math, you can't just convert nanoseconds to a pretty-print date without knowing the start date and end date and correctly calculating the days. That amount of nanoseconds could be a sliding scale and doesn't necessarily amount to a 22 year gap. For example, another 22 year gap is `2044-02-28 - 2022-02-28 | into duration --convert day` which is `8035` days, as is `1922-02-28 - 1900-02-28 | into duration --convert day`.

So, in order for date math to work in nushell, I think we'd need more than nanoseconds. We'd probably need the start date and end date to properly do date math. We could make that a new `Value::DateMath` or `Value::DateDuration` or some such value.

I'd love for this thesis to be proven wrong, so feel free to challenge my position.

**Comment by bobhy** [comment]
It's worth looking a bit at what commercial databases do for date arithmetic.  They are a mature environments where real money is at risk from incorrect date arithmetic, so they provide tools that make it hard/impossible to even express an incorrect calculation, and they represent an equilibrium that's been negotiated for a long time.

I'm referring to MS SQL Server below, because that's what I know, but these functions are part of the semi-standard SQL Language industry spec and not proprietary.

In SQL Server, datetime values are fine grained (10 ns, IIRC).  Duration values are ints with an associated unit of measure (ns, ms ... day, week, month and quarter).  
There is a [DATEDIFF](https://learn.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-ver16) function which returns a duration between 2 datetime values.  Caller must always specify the units in which the duration should be returned, and the result ignores the lower order bits of the respective datetimes.  The function is documented to count the number of duration unit boundaries crossed, so if you ask for the number of days between datetimes, the (hh:mm:ss.nnn) are ignored and the duration will always be a whole number of days. If you request a number of hours, it won't matter what the mm:ss portions of the datetimes are.  This implicit truncation is what prevents calculations from accumulating roundoff error.
There is a [DATEADD](https://learn.microsoft.com/en-us/sql/t-sql/functions/dateadd-transact-sql?view=sql-server-ver16) function which combines a datetime with a (positive or negative) duration which again always includes time span unit, and the result preserves whatever low order bits were in the input datetime.
 And those (and a few related functions with duration+unit in the signature) are the *only* ways you can do interval calculations in SQL. 
Translating these concepts into Nushell would involve removing support for datetime and duration types from arithmetic operators (because, e.g, you can't specify desired units in `<date> - <date>`) and adding subcommands to do the calculation in a carefully controlled way.

**Comment by fdncred** [comment]
Your datediff description is close to what's happening in this command
```
2022-02-28 - 2000-02-28 | into duration --convert day
```
it's converting to days, the underlying crate, chrono, calculates all the date math correctly, and the answer it gives is correct. The problem with nushell comes into play when the duration of nanoseconds get re-represented as human readable.

It sounds like you're understanding what I'm saying, so please forgive me repeating myself above.

I'm not sure we're willing to remove date math altogether from the repl but we are up for improving it. I'm definitely open to suggestions. For me, it's about having a better way to translate the duration to human readable values since the calculation is already correct but maybe I'm wrong?

**Comment by bobhy** [comment]
Um, the calculation *doesn't* work if the datetime values have sub-day precision:
```
〉("2000-03-02t01:00" | into datetime) - ("2000-03-01t18:00" | into datetime) | into duration --convert day
0.29 day
```
If you're running a hotel and charging by the nights, you'd want the above result to be 1, not rounded down to zero.
If you're calculating batch job run time, you might not care, though you might prefer to have the result reported in smaller time units.  

The only way to salvage the result using arithmetic operators would be to first truncate each of the terms to the desired duration unit, so you need a way to specify the duration unit to the operator, maybe something like:
```
〉2000-03-02t01:00 -day 2000-03-01t18:00 
```
Which looks a little weird (and the bare datetimes would not parse today), but maybe not worse than:
```
〉date days-between 2000-03-01t18:00 2000-03-02t01:00
```

**Comment by fdncred** [comment]
I'm not sure what you're getting at. If you want sub-day precision there's no need to convert to days. You'd do something like this.
```
> 2000-03-02T18:00:00 - 2000-03-02T01:00:00
17hr
```
or even more granular
```
> 2000-03-02T18:00:00 - 2000-03-02T01:25:12
16hr 34min 48sec
```

We could definitely have a `date math` command that takes two input datetimes, an operator, and an output unit (like days). But I really like the simplicity of just doing date math by putting datetimes in the repl. It's just so convenient.

**Comment by bobhy** [comment]
My point is that if I'm doing "financial" calendar calculations, the elegant date arithmetic operations on dates and durations don't work for me, and it's not clear how I should adapt them.
We talked about the hotel example above, where it's not the number of hours' duration, it's whether the duration includes midnight.  
Here's another, statistical, scenario. How do I group by date over a bunch of datetimes? That is, count the number of events per day, e.g, so I can plot my COVID infection chart?   The traditional solution is to truncate the datetime to a date, then aggregate by exact match on that value.   But the only way I seeto do that in Nushell is the hack of converting it to a string, there doesn't seem to be a `--convert ` like switch on `into datetime` like there is on `into` duration.  (But I'm a n00b, maybe I'm wrong about this one).

I take back my misguided suggestion to drop support for date and duration arithmetic with plus and minus.  That's probably too ingrained in people's fingertips to change now.  But it's a trap for the unwary enterprise developer, I do think we need extensions to support the other kind of calculation.
That might be  for which the extra precision Nushell currently supports is actually a bug. For these applications, the user really does want to know just the integer number of days (or maybe months...) between events that have a full-precision timestamp.  Doing that calculation so the books balance

For the hotel example I gave above, and many others you are familar with, but also for mortgages, medical billing , rHotels bill you for the number of midnights you were occupying the room (usually), even though you check in at, e.g, 1800 and check out the next morning at 0800.  When you close on a home purchase and the closing is in the middle of the month, the interest is prorated by some whole number of days.

**Comment by bobhy** [comment]
Hmm, I do think I could do something to  help on interval arithmetic in both  statistics and bookkeeping uses.  Will get on discord to see what flies.

**Comment by fdncred** [comment]
> Hmm, I do think I could do something to help on interval arithmetic in both statistics and bookkeeping uses. Will get on discord to see what flies.

I'm anxious to see what you come up with. :)

**Comment by rgwood** [comment]
This bit me again today. I wanted to view all files modified in the last year: `ls | where modified > ((date now) - 1yr)`

In this situation I did not care much about precision and I would have been fine with any sane definition of "year".

**Comment by fdncred** [comment]
> This bit me again today.

Agreed. It happens to me on occasion too. I've changed this so many times, researched it so much that I'm kind of burnt out on it not working the way I want it to. I think the easiest fix would be to allow yr and month as f64s like.
```rust
const DAYS_PER_YEAR: f64 = 365.2422; // so many opinions https://www.universityherald.com/articles/67348/20170301/the-most-precise-number-of-days-in-a-year-according-to-scientists.htm
const DAYS_PER_MONTH: f64 = DAYS_PER_YEAR / 12; // 30.43685
const DAYS_PER_DECADE: f64 = DAYS_PER_YEAR * 10; // 3652.422
```
Depending on how you want to calculate what "right" means, you'll get different answers, but if we were just to use something like this above, and document how we're getting to those numbers, at least we'd have something more helpful.

**Comment by cycomanic** [comment]
Let me join the discussion as a complete newbie with nushell. I recently read about nushell and wanted to try it out. 

So to get myself familiar with the way "piping" works in nushell I played around with the examples from the book in particular 
```
ls | where modified >  (date now) - 2wk
```

If I do `ls` in the same directory it also shows me lots of files as "2 months ago" ...  so naturally I tried lots of ways of trying writing  year or months , but always get type mismatch operators. 

In light of the previous discussion here are  2 cents. 

I realise that correct date math is hard (I guess that's what you get from a system build on centuries of arbitrary decisions and adjustments). However, `ls` already does date maths. Now I know that those age indications are approximate, but as a newbie I kind of expected that I would be able somehow replicate that behaviour (even if only approximate), i.e. select files older than 2 month. Not being able to do that was certainly confusing and let me to believe I'm doing something wrong (in particular I did not find any answer in the manual why this was not working). 

As I said, I understand solving this properly might not be easy, however in the interim I would suggest that `ls` should shows dates as absolute dates (which would also exact date calculations when comparing to the output of `ls`). At least there should be a note in the documentation about the fact that there durations cannot be specified in anything longer than weeks. 

Final note, I enjoy playing around with nushell, I think there's some really great new concepts here, which make shells much more fun to work with.

**Comment by fdncred** [comment]
To quote Michael Jackson, who wants to "make that change"? 😆

**Comment by bobhy** [comment]
Checking in from the island of Bali... Not me, not for the next 2 weeks.

Regards,
Bob
